File: src/141.linked-list-cycle/solution_hash-table.cpp
/*
 * @lc app=leetcode.cn id=141 lang=cpp
 *
 * https://leetcode-cn.com/problems/linked-list-cycle/
 */

#include <unordered_set>

using namespace std;

/**
 * Definition for singly-linked list.
 */
struct ListNode {
  int val;
  ListNode *next;
  ListNode(int x) : val(x), next(nullptr) {}
};

// @lc code=start

class Solution {
 public:
  bool hasCycle(ListNode *head) {
    unordered_set<ListNode *> dict;
    ListNode *curr = head;

    while (curr) {
      if (dict.find(curr) != dict.end()) {
        return true;
      }

      dict.insert(curr);
      curr = curr->next;
    }

    return false;
  }
};

// @lc code=end

File: src/141.linked-list-cycle/solution_hash-table.ts
/*
 * @lc app=leetcode.cn id=141 lang=typescript
 *
 * [141] 环形链表
 */

interface ListNode {
  val: number;
  next: ListNode | null;
}

// @lc code=start
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function hasCycle(head: ListNode | null): boolean {
  const set = new Set<ListNode>();
  let current = head;

  while (current) {
    if (set.has(current)) {
      return true;
    }

    set.add(current);
    current = current.next;
  }

  return false;
}
// @lc code=end

export default hasCycle;
